pkg/distributor/query.go

```diff
@@ -184,6 +211,81 @@ func (d *Distributor) queryIngesters(ctx context.Context, replicationSet ring.Re
 	return result, nil
 }
 
+// mergeExemplarSets merges and dedupes two sets of already sorted sample pairs.
+// Defined here instead of pkg/util to avoid a import cycle.
+func mergeExemplarSets(a, b []cortexpb.Exemplar) []cortexpb.Exemplar {
+	result := make([]cortexpb.Exemplar, 0, len(a)+len(b))
+	i, j := 0, 0
+	for i < len(a) && j < len(b) {
+		if a[i].TimestampMs < b[j].TimestampMs {
+			result = append(result, a[i])
+			i++
+		} else if a[i].TimestampMs > b[j].TimestampMs {
+			result = append(result, b[j])
+			j++
+		} else {
+			result = append(result, a[i])
+			i++
+			j++
+		}
+	}
+	// Add the rest of a or b. One of them is empty now.
+	result = append(result, a[i:]...)
+	result = append(result, b[j:]...)
+	return result
+}
+
+// queryIngestersExemplars queries the ingesters for exemplars.
+func (d *Distributor) queryIngestersExemplars(ctx context.Context, replicationSet ring.ReplicationSet, req *ingester_client.ExemplarQueryRequest) (*ingester_client.QueryResponse, error) {
+	// Fetch exemplars from multiple ingesters in parallel, using the replicationSet
+	// to deal with consistency.
+	results, err := replicationSet.Do(ctx, d.cfg.ExtraQueryDelay, func(ctx context.Context, ing *ring.InstanceDesc) (interface{}, error) {
+		client, err := d.ingesterPool.GetClientFor(ing.Addr)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := client.(ingester_client.IngesterClient).QueryExemplars(ctx, req)
+		d.ingesterQueries.WithLabelValues(ing.Addr).Inc()
+		if err != nil {
+			d.ingesterQueryFailures.WithLabelValues(ing.Addr).Inc()
+			return nil, err
+		}
+
+		return resp, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Merge results from replication set.
+	var keys []string
+	exemplarResults := make(map[string]cortexpb.TimeSeries)
+	for _, result := range results {
+		r := result.(*ingester_client.QueryResponse)
+		for _, ts := range r.Timeseries {
+			lbls := cortexpb.FromLabelAdaptersToLabels(ts.Labels).String()
```

Comment:
[non blocking] I'm wondering performance impact of this. Have you tried writing a benchmark and comparing it with using `Hash()` for the key?  The `keys []string` could then be a `keys []struct{ labels ..., hash ...}`